cookbook/core/extend_flyte/external_plugin_service.py

```diff
@@ -0,0 +1,119 @@
+
+Currently, flyteplugins is live in the flytepropeller, which means we have to rebuild flytepropeller if we register a new plugin.
+ It is hard to implement backend plugins, especially for data scientists & MLEs who do not have working knowledge of Golang.
+  Also, performance requirements, maintenance, and development are cumbersome.
+
+
+We built a new component in Flyte called the external-plugin-system. Some goals behind this,,
+
+* Plugins should be easy to author - no need for code generation, using tools that MLEs and Data Scientists are not accustomed to using.
+* The most important plugins for Flyte today are plugins that communicate with external services.
+* It should be possible to test these plugins independently and also deploy them privately.
+* It should be possible for users to use backend plugins for local development, especially in flytekit and unionML.
+* It should be possible to author plugins in any language.
+* Plugins should be scalable
+* Plugins should have a very simple API
+* Plugins should show up in the UI (extra details)
+
+Overview
+========
+The External Plugin System is a Python-based plugin registry that uses a gRPC server. Users and Propeller can send gRPC requests to this registry to run jobs like BigQuery and Databricks. Furthermore, the registry is stateless, which makes it easy to scale the system up or down.
+
+.. figure:: https://i.ibb.co/y0MhBfn/Screen-Shot-2023-04-16-at-11-51-17-PM.png
+  :alt: External-Plugin-System
+  :class: with-shadow
+
+Register a new plugin
+=====================
+
+Flytekit Interface Specification
+--------------------------------
+New backend plugins must be registered by extending the BackendPluginBase class in the flytekit backend module. All three methods below must be implemented, and all calls must be idempotent.
+
+- `create` is used to launch a new task. People can use gRPC, REST, or SDK to create a task.
+- `get` is used to get the jobID (like BigQuery Job ID or Databricks task ID).
+- `delete` Will send a request to delete the job.
+
+.. code-block:: python
+
+    class BackendPluginBase:
+    def __init__(self, task_type: str):
+        self._task_type = task_type
+
+    @property
+    def task_type(self) -> str:
+        return self._task_type
+
+    @abstractmethod
+    def create(
+        self,
+        context: grpc.ServicerContext,
+        output_prefix: str,
+        task_template: TaskTemplate,
+        inputs: typing.Optional[LiteralMap] = None,
+    ) -> TaskCreateResponse:
+        pass
+
+    @abstractmethod
+    def get(self, context: grpc.ServicerContext, job_id: str) -> TaskGetResponse:
+        pass
+
+    @abstractmethod
+    def delete(self, context: grpc.ServicerContext, job_id: str) -> TaskDeleteResponse:
+        pass
+
+    # To register the custom plugin
+    BackendPluginRegistry.register(CustomPlugin())
+
+Build a New image
+-----------------
+The following is a sample Dockerfile for building an image for an external plugin system.
```

Comment:
Can you explain why this needs to be done?

Reply:
we run a gRPC server (external plugin service) in kubernetes. propeller sends the request (create / delete / get task) to this gRPC server. external plugin service is a deployment in the kubernetes, so we should build an image for it.